ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionHeartbeatService.java

```diff
@@ -0,0 +1,213 @@
+    initialDelay = txnTimeout / 4;
+    period = txnTimeout / 2;
+  }
+
+  private final class CompactionHeartbeater {
+    private final Logger LOG = LoggerFactory.getLogger(CompactionHeartbeater.class);
+    private final long txnId;
+    private final long lockId;
+    private final String tableName;
+    private final ScheduledThreadPoolExecutor heartbeatExecutor;
+
+    private CompactionHeartbeater(long txnId, long lockId, String tableName) {
+      heartbeatExecutor = new ScheduledThreadPoolExecutor(1);
+      this.tableName = Objects.requireNonNull(tableName);
+      this.txnId = txnId;
+      this.lockId = lockId;
+    }
+
+    void start() {
+      heartbeatExecutor.scheduleAtFixedRate(() -> {
+        IMetaStoreClient msc = null;
+        try {
+          LOG.debug("Heartbeating compaction transaction id {} for table: {}", txnId, tableName);
+          // Create a metastore client for each thread since it is not thread safe
+          msc = clientPool.borrowObject();
+          msc.heartbeat(txnId, lockId);
+        } catch (NoSuchElementException e) {
+          LOG.error("Compaction transaction heartbeater pool exhausted, unable to heartbeat", e);
+          // This heartbeat attempt failed, and there is no client to return to the pool.
+          return;
+        } catch (TException e) {
+          LOG.error("Error while heartbeating compaction transaction id {} for table: {}", txnId, tableName, e);
+          // Heartbeat failed, but the client is not broken, we can return it to the pool.
+        } catch (Exception e) {
+          LOG.error("Error while heartbeating compaction transaction id {} for table: {}", txnId, tableName, e);
+          // Unknown error, invalidate the client, maybe it is broken.
+          if (msc != null) {
+            try {
+              clientPool.invalidateObject(msc);
+            } catch (Exception ex) {
+              LOG.error("Error while invalidating a broken MetaStoreClient instance", e);
+            }
+          }
+          return;
+        }
+        try {
+          if (msc != null) {
+            clientPool.returnObject(msc);
+          }
+        } catch (Exception e) {
+          LOG.error("Error while invalidating a broken MetaStoreClient instance", e);
+        }
+      }, initialDelay, period, TimeUnit.MILLISECONDS);
+    }
+
+    public void stop() {
+      LOG.info("Shutting down compaction txn heartbeater instance.");
+      heartbeatExecutor.shutdownNow();
+      try {
+        if (!heartbeatExecutor.awaitTermination(5, TimeUnit.SECONDS)) {
+          LOG.warn("Heartbeating for transaction {} did not stop in 5 seconds, do not wait any longer.", this);
+          return;
+        }
+      } catch (InterruptedException ex) {
+        //Caller thread was interrupted while waiting for heartbeater to terminate.
+        //Nothing to do, just restore the interrupted state.
+        Thread.currentThread().interrupt();
```

Comment:
how would it behave in case shutdownhook is invoked? would it interrupt the shutdownhook thread?